### DISCUSSION TITLE
Existe algum "positionSheetByIndex(const aSheetIndex: integer)"

### DISCUSSION THREAD

**Discussion by adolfomayer** [Daniel09Fernandes/Component_OpenOffice#62] [Q&A] [answered]
Olá a todos.
Tive uma necessidade aqui que talvez outras pessoas possam ter e não sei se tem implementação. Trata-se de um metodo que poderia ser  "positionSheetByIndex(const aSheetIndex: integer)" pois no documento que eu precisei interagir, havia uma planilha com um nome aleatório e uma segunda planilha com o nome "Planilha 1". Como não encontrei esse método ou algo parecido para posicionar na planilha de indice(0), independente do nome, fiquei sem saber como resolver. Alguém tem alguma ideia?

**Comment by Daniel09Fernandes** [comment, marked as answer]
@adolfomayer  Subi a alteração sugerida, pode utilizar o positionSheetByIndex(const aSheetIndex: integer), para navegar entre planilhas por Index e não só pelo nome. basta fazer o PULL para obter a nova atualização.

**Reply by adolfomayer** [reply]
Massa, Daniel. 
Eu ontem acabei criando esse método e mais dois que precisei. Um para deletar linhas e outro para deletar colunas (era uma necessidade iminente e eu improvisei). Não conheço muito esse universo de métodos, classe, objetos. Sou acostumado ao feijão com arroz do Delphi. Então pesquisei bastante nos forums do OpenOffice e vi algumas coisas que poderia tentar usar. Não sei se implementei no lugar correto, nem se foi da melhor maneira, mas depois de quebrar a cabeça aqui, acabou funcionando.. Vou olhar as diferenças no seu código pra ajustar aqui. Valeu demais! Foi muito bom encontrar esse seu projeto, porque eu uso Delphi e esyou interagindo com planilhas do LibreOffice. Parabéns!

**Reply by Daniel09Fernandes** [reply]
Muito obrigado @adolfomayer  fico feliz em ouvir seu feedback, se quiser contribuir com os métodos que criou de deletar linhas e colunas, pode subir um PR para o projeto, ou me passar que eu implemento aqui e deixo como oficial. 
Se precisar de algo, só me acionar, grande abraço.

**Reply by adolfomayer** [reply]
Oi Daniel.

Vou tentar organizar os métodos aqui pra te enviar oscódigos. Como não usei o git pra baixar o DinosOffice, vou mandar anexo por aqui. Só peço que dê um desconto na minha implementação porque, como te disse, não sei fazer direito essas coisas de classe e tudo o mais. Provavelmente não ficou da maneira correta, embora tenha funcionado pra mim, kkkk

Vou te fazer outra pergunta aqui mesmo, mas se você achar melhor colocar em como comentário, pra que outras pessoas possam ver, me diga que eu faço isso. É o seguinte: No meu aplicativo, eu abro um documento do MS Excel (formato .xlsx) e faço uma "higienização" na planilha. O que restar, preciso salvar em outro arquivo no mesmo formato (.xlsx) mas o OpenOffice não parece aceitar o formato.. 

Eu alterei o método SaveFile  na unit uOpenOffice pra ficar assim: 

procedure TOpenOffice.saveFile(aFileName: String; ExcelFormat: Boolean = false);
var
  SaveProperty : array [0..1] of variant;
begin
  SaveProperty[0] := ObjServiceManager.Bridge_GetStruct('com.sun.star.beans.PropertyValue');
  SaveProperty[0].Name := 'FilterName';

  { Tive que forçar aqui, porque o filtro "Excel 2007-365" que vemos quando vamos salvar
    manualmente  "file->Save as" não funciona e dá erro. Pelo que pesquisei é como se esse filtro não existisse para o OpenOffice}

  SaveProperty[0].Value := 'MS Excel 97'; 
  SaveProperty[1] := ObjServiceManager.Bridge_GetStruct('com.sun.star.beans.PropertyValue');
  SaveProperty[1].Name := 'Overwrite';
  SaveProperty[1].Value := True;

  aFileName := convertFilePathToUrlFile(aFileName);

  if aFileName.Trim.IsEmpty then
    aFileName := URlFile;
  if ExcelFormat then
    objDocument.storeAsURL(aFileName, VarArrayOf(SaveProperty))
  else
    objDocument.storeAsURL(aFileName, VarArrayOf([]));
end;

Vc já viu algo nesse sentido?
Abraços!

**Reply by Daniel09Fernandes** [reply]
Olá Adolfo, opa, toda contribuição é bem-vinda haha, obrigado meu amigo. 

Humm, não tinha me deparado com esse cenário ainda, com a aplicação desse filtro, vc conseguiu salvar em XLSX ? 

Se sim, vou aplicar, no código aqui e testar, se não, eu vejo como fazer aqui.

**Reply by Daniel09Fernandes** [reply]
@adolfomayer Teste aqui o código, ficou ótimo, fiz pequenas modificações e vou subir ele. 

```
procedure TOpenOffice.saveFile(aFileName: String);
var
  SaveProperty : array [0..1] of variant;
begin
  aFileName := convertFilePathToUrlFile(aFileName);

  if aFileName.Contains('.xlsx') then
  begin
    //Codigo fornecido por @adolfomayer - e adptado por @dinosdev 29/11/2024
    SaveProperty[0] := ObjServiceManager.Bridge_GetStruct('com.sun.star.beans.PropertyValue');
    SaveProperty[0].Name := 'FilterName';
    SaveProperty[0].Value := 'Calc MS Excel 2007 XML'; //for XLSX

    SaveProperty[1] := ObjServiceManager.Bridge_GetStruct('com.sun.star.beans.PropertyValue');
    SaveProperty[1].Name := 'Overwrite';
    SaveProperty[1].Value := True;
    objDocument.storeAsURL(aFileName, VarArrayOf(SaveProperty))
  end
  else
    objDocument.storeAsURL(aFileName, VarArrayOf([]));
end;
```
